src/esp32/src/flash.c

```diff
@@ -4,16 +4,148 @@
  * SPDX-License-Identifier: Apache-2.0 OR MIT
  */
 
-#include "flash.h"
+#include <target/flash.h>
+
+#include <stdbool.h>
+#include <stddef.h>
+
+#include <log.h>
+#include <private/rom.h>
+#include <private/soc_utils.h>
+
+#include "soc/gpio_reg.h"
+#include "soc/spi_reg.h"
+
+#define SPI_FLASH_CMD_RDID 0x9F
+#define ESP_ROM_SPIFLASH_BUSY_FLAG BIT(0)
+
+#define PERIPHS_SPI_FLASH_CMD SPI_CMD_REG(1)
+#define PERIPHS_SPI_FLASH_STATUS SPI_RD_STATUS_REG(1)
+#define PERIPHS_SPI_FLASH_USRREG SPI_USER_REG(1)
+#define PERIPHS_SPI_FLASH_USRREG2 SPI_USER2_REG(1)
+#define PERIPHS_SPI_FLASH_C0 SPI_W0_REG(1)
+#define PERIPHS_SPI_MOSI_DLEN_REG SPI_MOSI_DLEN_REG(1)
+#define PERIPHS_SPI_MISO_DLEN_REG SPI_MISO_DLEN_REG(1)
+#define SPI_USR2_DLEN_SHIFT SPI_USR_COMMAND_BITLEN_S
+
+/* ROM */
+
+extern uint32_t ets_efuse_get_spiconfig(void);
+extern void esp_rom_spiflash_attach(uint32_t ishspi, bool legacy);
+extern esp_rom_spiflash_chip_t g_rom_flashchip;
+extern uint8_t g_rom_spiflash_dummy_len_plus[];
 
 void stub_target_flash_init(void *state)
 {
+    STUB_LOG_TRACE();
     (void)state;
-    // TODO: Implement
+    uint32_t spiconfig = ets_efuse_get_spiconfig();
+
+    uint32_t strapping = REG_READ(GPIO_STRAP_REG);
+    /*  If GPIO1 (U0TXD) is pulled low and flash pin configuration is not set in efuse, assume
+     * HSPI flash mode (same as normal boot) */
+    if (spiconfig == 0 && (strapping & 0x1c) == 0x08) {
+        spiconfig = 1;    /* HSPI flash mode */
+    }
+
+    // TODO: do we need legacy arg = !ETS_IS_FAST_FLASH_BOOT() ?
+    esp_rom_spiflash_attach(spiconfig, false);
 }
 
 void stub_target_flash_deinit(const void *state)
 {
     (void)state;
     // TODO: Implement
 }
+
+void stub_target_flash_unlock(void)
+{
+    // TODO: Add the patched version to rom_common
+}
+
+// TODO: common code with esp32_flash_spi_wait_ready
+static uint32_t esp32_flash_exec_usr_cmd(uint32_t cmd)
```

Comment:
Since we are in the esp32 directory, we don't need `esp32_` prefixes in the function names. It would be nice to follow some naming convention also.  For example start with `flash_ ` if they are not public.